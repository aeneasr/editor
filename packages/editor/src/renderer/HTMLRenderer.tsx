import classNames from 'classnames';
import React from 'react';
import { migrateValue } from '../core/migrations/migrate';
import { setAllSizesAndOptimize } from '../core/reducer/value/helper/setAllSizesAndOptimize';
import { optimizeRows } from '../core/reducer/value/helper/optimize';
import type { Cell, CellPlugin, Row, ValueWithLegacy } from '../core/types';
import { getCellData } from '../core/utils/getCellData';
import { getCellStyle } from '../core/utils/getCellStyle';
import {
  getPluginCellSpacing,
  normalizeCellSpacing,
} from '../core/utils/getCellSpacing';

const gridClass = (size = 12): string =>
  `react-page-cell-sm-${size} react-page-cell-xs-12`;

const rowHasInlineChildren = ({ cells }) =>
  Boolean(cells.length === 2 && Boolean(cells[0].inline));

const HTMLRow: React.FC<
  Partial<
    Row & {
      lang: string;
      className?: string;
      cellPlugins: CellPlugin[];
      cellSpacing: [number, number];
    }
  >
> = React.memo(
  ({
    cells = [],
    className,
    lang,
    cellPlugins,
    cellSpacing: [cellSpacingX, cellSpacingY],
  }) => (
    <div
      className={classNames('react-page-row', className, {
        'react-page-row-has-floating-children': rowHasInlineChildren({ cells }),
      })}
      style={{ margin: `0 ${-cellSpacingX / 2}px` }}
    >
      {cells.map((c, index) => (
        <HTMLCell
          key={c.id}
          {...c}
          lang={lang}
          cellPlugins={cellPlugins}
          cellSpacing={[cellSpacingX, cellSpacingY]}
        />
      ))}
    </div>
  )
);

// eslint-disable-next-line no-empty-function
const noop = () => {
  return;
};

const DefaultProvider: React.FC = ({ children }) => <>{children}</>;

const HTMLCell: React.FC<
  Cell & {
    lang: string;
    cellPlugins: CellPlugin[];
    cellSpacing: [number, number];
  }
> = React.memo((props) => {
  const {
    lang,
    cellPlugins,
    cellSpacing: [cellSpacingX, cellSpacingY],
    ...cell
  } = props;
  const { size, hasInlineNeighbour, inline, isDraftI18n, isDraft } = cell;
  const hasChildren = cell.rows?.length > 0;
  const cnOuter = classNames(gridClass(size), {
    'react-page-cell-has-inline-neighbour': hasInlineNeighbour,
    [`react-page-cell-inline-${inline || ''}`]: inline,
  });
  const cnInner = classNames('react-page-cell', {
    'react-page-cell-leaf': !hasChildren,
  });

  if (isDraftI18n?.[lang] ?? isDraft) {
    return null;
  }

  const plugin = cell.plugin
    ? cellPlugins.find((p) => p.id === cell.plugin.id)
    : null;
  if (plugin) {
    const { Renderer } = plugin;
    const cellStyle = getCellStyle(plugin);
    const Provider = plugin.Provider ?? DefaultProvider;
    const data = getCellData(cell, lang);
    const pluginCellSpacing = getPluginCellSpacing(plugin, data);
    const [pluginCellSpacingX, pluginCellSpacingY] = pluginCellSpacing
      ? normalizeCellSpacing(pluginCellSpacing)
      : [cellSpacingX, cellSpacingY];
    const props = {
      readOnly: true,
      lang: lang,
      nodeId: cell.id,
      data: data,
      onChange: noop,
      pluginConfig: plugin,
      focused: false,
      isPreviewMode: false,
      isEditMode: false,
    };
    return (
      <Provider {...props}>
        <div
          className={cnOuter}
          style={{ padding: `${cellSpacingY / 2}px ${cellSpacingX / 2}px` }}
        >
          <div className={cnInner}>
            <div
              style={cellStyle}
              className={
                'react-page-cell-inner' +
                (cell.rows?.length > 0 ? '' : ' react-page-cell-inner-leaf')
              }
            >
              <div
                style={{
                  display: 'flex',
                  flexDirection: 'column',
                }}
              >
                <Renderer {...props}>
                  {cell.rows?.length ? (
                    <div style={{ margin: `${-pluginCellSpacingY / 2}px 0` }}>
                      {cell.rows?.map((r: Row) => (
                        <HTMLRow
                          key={r.id}
                          {...r}
                          cellPlugins={cellPlugins}
                          cellSpacing={[pluginCellSpacingX, pluginCellSpacingY]}
                          lang={lang}
                          className="react-page-cell-inner"
                        />
                      ))}
                    </div>
                  ) : null}
                </Renderer>
              </div>
            </div>
          </div>
        </div>
      </Provider>
    );
  } else if (cell.rows?.length > 0) {
    return (
      <div className={cnOuter} style={{ padding: `0 ${cellSpacingX / 2}px` }}>
        <div className={cnInner}>
          {cell.rows.map((r: Row) => (
            <HTMLRow
              key={r.id}
              {...r}
              lang={lang}
              className="react-page-cell-inner"
              cellPlugins={cellPlugins}
              cellSpacing={[cellSpacingX, cellSpacingY]}
            />
          ))}
        </div>
      </div>
    );
  }

  return (
    <div className={cnOuter}>
      <div className={cnInner}>
        <div className="react-page-cell-inner" />
      </div>
    </div>
  );
});

export interface HTMLRendererProps {
  value: ValueWithLegacy;
  cellPlugins?: CellPlugin[];
  lang?: string;
  cellSpacing?: number | [number, number];
}

export const HTMLRenderer: React.FC<HTMLRendererProps> = React.memo(
  ({ value, cellPlugins, cellSpacing, lang = 'default' }) => {
    const data = migrateValue(value, { cellPlugins, lang });
    const [cellSpacingX, cellSpacingY] = normalizeCellSpacing(cellSpacing);

    if (!data) {
      return null;
    }
    const { rows } = data;
    const optRows = optimizeRows(rows);
    return (
      <div
        style={{ margin: optRows?.length ? `${-cellSpacingY / 2}px 0` : null }}
      >
        {setAllSizesAndOptimize(optRows).map((row) => (
          <HTMLRow
            key={row.id}
            cellPlugins={cellPlugins}
            lang={lang}
            cellSpacing={[cellSpacingX, cellSpacingY]}
            {...row}
          />
        ))}
      </div>
    );
  }
);
