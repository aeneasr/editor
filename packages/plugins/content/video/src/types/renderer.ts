import type { VideoProps } from './component';

export type VideoHtmlRendererProps = VideoProps;
