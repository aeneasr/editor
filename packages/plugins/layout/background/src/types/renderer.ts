import { BackgroundProps } from './component';
import { RGBColor } from 'ory-editor-ui/lib/ColorPicker';
import { ImageLoaded } from 'ory-editor-ui/lib/ImageUpload';

export interface BackgroundRendererExtraProps {
  backgroundColorPreview?: RGBColor;
  gradientDegPreview?: number;
  gradientDegPreviewIndex?: number;
  gradientOpacityPreview?: number;
  gradientOpacityPreviewIndex?: number;
  gradientColorPreview?: RGBColor;
  gradientColorPreviewIndex?: number;
  gradientColorPreviewColorIndex?: number;
  darkenPreview?: number;
  lightenPreview?: number;
  imagePreview?: ImageLoaded;
  forecolorPreview?: string;
  backgroundSize?: string;
  backgroundPositionH?: string;
  backgroundPositionV?: string;
}

export type BackgroundRendererProps = BackgroundProps & BackgroundRendererExtraProps;