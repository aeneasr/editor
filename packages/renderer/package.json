{
  "name": "ory-editor-renderer",
  "version": "0.0.0",
  "main": "./lib/index.js",
  "author": "ORY GmbH",
  "license": "AGPL-3.0",
  "scripts": {
    "build": "npm-run-all --parallel build:lib",
    "build:lib": "babel ./src --ignore test.js --out-dir ./lib",
    "build:lib:watch": "npm run build:lib -- --watch",
    "build:css": "postcss -c .postcss.json -o lib/styles.css src/*.css",
    "build:css:watch": "npm run watch:js -- -c \"npm run build:css\"",
    "clean": "rimraf \"dist/*\" \"lib/*\"",
    "coveralls": "cat ./coverage/lcov.info | ./node_modules/coveralls/bin/coveralls.js",
    "docs": "documentation build ./src/**/*.js --format md --github -o ../../docs/api/renderer.md",
    "flow": "flow check --show-all-errors",
    "flow:watch": "npm run watch:js -- -c \"flow check --show-all-errors\"",
    "lint": "eslint \"+(src)/**/*.js\" --fix",
    "lint:watch": "npm run watch:js -- -c \"npm run lint\"",
    "prebuild": "npm run clean",
    "start": "npm-run-all --parallel build:lib:watch build:css:watch",
    "test": "jest --coverage",
    "test:watch": "jest --watch",
    "watch:js": "chokidar \"+(src)/**/*.js\" --initial"
  },
  "peerDependencies": {
    "react": "^15.4.2",
    "react-dom": "^15.4.2"
  },
  "dependencies": {
    "classnames": "^2.2.5",
    "ory-editor-core": "0.0.0"
  }
}
