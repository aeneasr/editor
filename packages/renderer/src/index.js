// @flow
import React from 'react'
import classNames from 'classnames'
import PluginService from 'ory-editor-core/lib/service/plugin'
import { editable as reducer }from 'ory-editor-core/lib/reducer/editable'

const gridClass = (size = 12): string => {
  return `ory-cell-md-${size} ory-cell-xs-12`
}

const HTMLRow = ({ cells = [], className, hasInlineChildren }) => (
  <div className={classNames('ory-row', className, {
  'ory-row-has-floating-children': hasInlineChildren
  })}>
    {cells.map((c) => <HTMLCell key={c.id} {...c} />)}
  </div>
)


const HTMLCell = (props) => {
  const { rows = [], layout = {}, content = {}, hasInlineNeighbour, inline, size } = props
  const cn = classNames('ory-cell', gridClass(size), {
    'ory-cell-has-inline-neighbour': hasInlineNeighbour,
    [`ory-cell-inline-${inline || ''}`]: inline
  })

  if (layout.plugin) {
    const { state, plugin: { Component } } = layout
    return (
      <div className={cn}>
        <div className="ory-cell-inner">
          <Component
            readOnly={true}
            state={state}
          >
            {rows.map(r => <HTMLRow key={r.id} {...r} className="ory-cell-inner"/>)}
          </Component>
        </div>
      </div>
    )
  } else if (content.plugin) {
    const { state, plugin: { Component } } = content
    return (
      <div className={cn}>
        <div className="ory-cell-inner ory-cell-leaf">
          <Component
            readOnly={true}
            state={state}
          />
        </div>
      </div>
    )
  } else if (rows.length > 0) {
    return (
      <div className={cn}>
        {rows.map(r => <HTMLRow key={r.id} {...r} className="ory-cell-inner"/>)}
      </div>
    )
  }

  return (
    <div className={cn}>
      <div className="ory-cell-inner">empty</div>
    </div>
  )
}

export const HTMLRenderer = ({ state, plugins }) => {
  const service = new PluginService(plugins)
  const props = reducer(service.unserialize(state), { type: 'renderer/noop' })
  return <HTMLRow {...props}/>
}
