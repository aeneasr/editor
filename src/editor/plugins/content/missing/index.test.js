/* eslint-env jest */
import React from 'react'
import renderer from 'react-test-renderer'

import MissingPlugin from './index'

describe('plugins/content/missing', () => {
  it('should render', () => {
    const { Component } = MissingPlugin
    const component = renderer.create(<Component name="Foo" version="0.0.1" />)
    expect(component).toMatchSnapshot()
  })
})
