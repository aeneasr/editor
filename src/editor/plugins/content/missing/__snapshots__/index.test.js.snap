exports[`plugins/content/missing should render 1`] = `
<div
  style={
    Object {
      "backgroundColor": "red",
    }
  }>
  Plugin 
  <code>
    Foo
    :
    0.0.1
  </code>
   not found.
</div>
`;
